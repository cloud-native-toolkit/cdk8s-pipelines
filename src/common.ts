const paramNameRegex = new RegExp('^[a-zA-Z_][-a-zA-Z0-9_]+$');

function hasValidName(res: NamedResource): boolean {
  let valid = false;
  if (res.name) {
    valid = paramNameRegex.test(res.name);
  }
  return valid;
}

export interface NamedResource {
  readonly name?: string;
}
