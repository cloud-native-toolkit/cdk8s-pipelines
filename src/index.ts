export * from './pipelines';
export * from './tasks';
export * from './common';
