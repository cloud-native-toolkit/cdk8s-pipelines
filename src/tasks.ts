/**
 * This file includes the basic objects for creating Tasks and dependencies.
 * While cdk8s supports generating this file from the CRD, the Tekton CRD at
 * the moment does not have the fields specified. See https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml,
 * specfically here:
 *       schema:
 *         openAPIV3Schema:
 *           type: object
 *           # OpenAPIV3 schema allows Kubernetes to perform validation on the schema fields
 *           # and use the schema in tooling such as `kubectl explain`.
 *           # Using "x-kubernetes-preserve-unknown-fields: true"
 *           # at the root of the schema (or within it) allows arbitrary fields.
 *           # We currently perform our own validation separately.
 *           # See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema
 *           # for more info.
 *           x-kubernetes-preserve-unknown-fields: true
 * For the time being, these objects are created manually using the latest
 * Tekton documentation as a reference.
 */
import { ApiObject, ApiObjectMetadata, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';
import { NamedResource } from './common';

/**
 * A workspace used by a Task. See https://tekton.dev/docs/pipelines/workspaces/#using-workspaces-in-tasks for more information.
 */
export interface TaskWorkspace extends NamedResource {
  readonly workspace?: string;
}

export interface TaskWorkspaceRef extends TaskWorkspace {
  readonly description?: string;
}

export class TaskRef {
  name?: string;

  constructor(name: string) {
    this.name = name;
  }
}

/**
 * A Task parameter value.
 */
export interface TaskParam extends NamedResource {
  /**
   * The value of the task parameter.
   */
  readonly value?: string;
}

/**
 * Specifies execution parameters for the Task.
 */
export interface TaskSpecParam extends NamedResource {
  readonly type?: string;
  readonly description?: string;
  readonly default?: string;
}

/**
 * The volume mount for the task.
 */
export interface TaskVolumeMount extends NamedResource {
  readonly mountPath?: string;
}

/**
 * The step for a Task. See https://tekton.dev/docs/pipelines/tasks/#defining-steps
 * @see https://tekton.dev/docs/pipelines/container-contract/
 */
export interface TaskStep extends NamedResource {
  /**
   * The name of the container image to use for the Step.
   */
  readonly image?: string;
  /**
   * A script that will be executed on the image. If `script` is specified,
   * then the `command` value should not be specified.
   */
  readonly script?: string;
  /**
   * The command and its arguments (provided in the form of an array) to
   * execute on the container. If `command` is supplied, you should not supply
   * `script`.
   */
  readonly command?: string[];
  /**
   * Alternatively, you can supply `args` to the `command` value here.
   */
  readonly args?: string[];
  /**
   * The volume mounts to use for the task.
   */
  readonly volumeMounts?: TaskVolumeMount[];
  /**
   * The name of the working directory for the step
   */
  readonly workingDir?: string;
}

/**
 * The Task spec.
 *
 * @see https://tekton.dev/docs/pipelines/tasks/#configuring-a-task
 */
export interface TaskSpec {
  /**
   * The description of the `Task`.
   * @see https://tekton.dev/docs/pipelines/tasks/#adding-a-description
   */
  readonly description?: string;
  /**
   * The `Task`'s parameters.
   * @see https://tekton.dev/docs/pipelines/tasks/#specifying-parameters
   */
  readonly params?: TaskSpecParam[];
  /**
   * The steps that will be executed as part of the Task. The `Step` should
   * fit the (container contract)[https://tekton.dev/docs/pipelines/container-contract/]
   * @see https://tekton.dev/docs/pipelines/tasks/#defining-steps
   */
  readonly steps?: TaskStep[];
}

/**
 * Properties used to create the Task.
 */
export interface TaskProps {
  /**
   * The object [metadata](https://kubernetes.io/docs/concepts/overview/working-with-objects/#required-fields)
   * that conforms to standard Kubernetes metadata.
   */
  readonly metadata?: ApiObjectMetadata;
  /**
   * The `spec` is the configuration of the `Task` object.
   */
  readonly spec?: TaskSpec;
}

/**
 * A Tekton Task, which is
 * > a collection of Steps that you define and arrange in
 * > a specific order of execution as part of your continuous integration flow. A
 * > Task executes as a Pod on your Kubernetes cluster. A Task is available within a
 * > specific namespace, while a ClusterTask is available across the entire
 * > cluster.
 *
 * @see https://tekton.dev/docs/pipelines/tasks/
 */
export class Task extends ApiObject {
  /**
   * Returns the apiVersion and kind for "Task"
   */
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'tekton.dev/v1beta1',
    kind: 'Task',
  };

  /**
   * Renders a Kubernetes manifest for "Task".
   *
   * This can be used to inline resource manifests inside other objects (e.g. as templates).
   *
   * @param props initialization props
   */
  public static manifest(props: TaskProps = {}): any {
    return {
      ...Task.GVK,
      ...props,
    };
  }

  private readonly _metadata?: ApiObjectMetadata;
  private readonly _spec?: TaskSpec;

  /**
   * Defines a "Task" API object
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialization props
   */
  public constructor(scope: Construct, id: string, props: TaskProps = {}) {
    super(scope, id, {
      ...Task.GVK,
      ...props,
    });
    this._metadata = props.metadata;
    this._spec = props.spec;
  }

  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const result = {
      ...Task.GVK,
      ...{
        metadata: this._metadata,
        spec: this._spec,
      },
    };
    return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({
      ...r,
      [i[0]]: i[1],
    }), {});
  }
}
