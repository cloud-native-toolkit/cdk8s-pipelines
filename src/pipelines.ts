import { ApiObject, ApiObjectMetadata, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';
import { NamedResource } from './common';
import { TaskParam, TaskRef, TaskWorkspace, TaskWorkspaceRef } from './tasks';

// The following interfaces and classes are strictly for generating the YAML or
// JSON in the proper format for the Tekton pipelines. See the builders to use
// classes and a fluid-style syntax that provides easier building and checking.

/**
 * A Pipeline parameter. See https://tekton.dev/docs/pipelines/pipelines/#specifying-parameters
 */
export interface PipelineParam extends NamedResource {
  readonly type?: string;
  readonly default?: string;
}

/**
 * A task in a pipeline. See https://tekton.dev/docs/pipelines/pipelines/#adding-tasks-to-the-pipeline
 */
export interface PipelineTask extends NamedResource {
  readonly taskRef?: TaskRef;
  readonly params?: TaskParam[];
  readonly runAfter?: string[];
  readonly workspaces?: TaskWorkspace[];
}

/**
 * A workspace for a pipeline. See https://tekton.dev/docs/pipelines/pipelines/#specifying-workspaces
 * and https://tekton.dev/docs/pipelines/workspaces/#using-workspaces-in-pipelines.
 */
export interface PipelineWorkspace extends NamedResource {
  /**
   * The description of the workspace.
   */
  readonly description?: string;
}

/**
 * A task definition at the pipeline level.
 */
export interface PipelineTaskDef extends PipelineTask {
  readonly refParams?: PipelineParam[];
  readonly refWorkspaces?: TaskWorkspaceRef[];
}

/**
 * The `spec` part of the Pipeline.
 *
 * @see https://tekton.dev/docs/pipelines/pipelines/
 */
export interface PipelineSpec {
  readonly description?: string;
  readonly params?: PipelineParam[];
  /**
   * The `tasks` are required on the Pipeline.
   */
  readonly tasks: PipelineTask[];
  readonly workspaces?: PipelineWorkspace[];
}

/**
 * Properties used to create the Pipelines.
 */
export interface PipelineProps {
  readonly metadata?: ApiObjectMetadata;
  readonly spec?: PipelineSpec;
}

/**
 *
 *
 * @schema Pipeline
 */
export class Pipeline extends ApiObject {

  /**
   * Returns the apiVersion and kind for "Pipeline"
   */
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'tekton.dev/v1beta1',
    kind: 'Pipeline',
  };

  private readonly _metadata?: ApiObjectMetadata;
  private readonly _spec?: PipelineSpec;

  /**
   * Renders a Kubernetes manifest for "Pipeline".
   *
   * This can be used to inline resource manifests inside other objects (e.g. as templates).
   *
   * @param props initialization props
   */
  // public static manifest(props: PipelineProps = {}): any {
  //   return {
  //     ...Pipeline.GVK,
  //     ...
  //   };
  // }

  /**
   * Defines a "Pipeline" API object
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialization props
   */
  public constructor(scope: Construct, id: string, props: PipelineProps = {}) {
    super(scope, id, {
      ...Pipeline.GVK,
    });
    this._metadata = props.metadata;
    this._spec = props.spec;
  }


  /**
   * Renders the object to Kubernetes JSON.
   */
  public toJson(): any {
    const result = {
      ...Pipeline.GVK,
      ...{
        metadata: this._metadata,
        spec: this._spec,
      },
    };
    return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({
      ...r,
      [i[0]]: i[1],
    }), {});
  }
}
