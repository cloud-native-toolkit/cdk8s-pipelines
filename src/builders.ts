/**
 * This file has builders in it for the various pipeline constructs.
 */


import { Pipeline, PipelineTaskDef } from './pipelines';
import { Construct } from 'constructs';
import { TaskRef } from './tasks';

/**
 * The PipelineTaskBuilder creates a PipelineTaskDef, which is used to
 * intelligently add a task and its properties to a Pipeline.
 */
export class PipelineTaskBuilder {
  private name?: string;
  private taskRefName?: string;
  /**
   * Creates a new instance of the PipelineTaskBuilder
   */
  constructor() {
  }

  public withName(name: string): PipelineTaskBuilder {
    this.name = name;
    return this;
  }

  public withTaskReference(ref: string): PipelineTaskBuilder {
    this.taskRefName = ref;
    return this;
  }

  /**
   * Builds the Pipeline
   */
  public build() : PipelineTaskDef {
    // TODO: An important part of build here is to assert that the
    // object has been correctly made and throw very meaningful errors.
    const pt = {
      name: this.name,
      taskRef: new TaskRef(this.taskRefName!),
      params: [
        {
          name: 'url',
          value: '$(params.repo-url)',
        },
      ],
      refParams: [
        {
          name: 'repo-url',
          default: '',
          type: 'string',
        },
      ],
      refWorkspaces: [
        {
          name: 'output',
          workspace: 'shared-files',
          description: 'The files that are cloned by the task',
        },
      ],
    };
    return pt;
  }
}

export class PipelineBuilder {
  constructor() {
  }
}