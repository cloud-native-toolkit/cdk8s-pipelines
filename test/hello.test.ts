import { Chart, Testing } from 'cdk8s';
import { ChartProps } from 'cdk8s/lib/chart';
import { Construct } from 'constructs';
// @ts-ignore
import { Pipeline, PipelineTaskBuilder, PipelineTaskDef, TaskRef } from '../src';

class MyTestChart extends Chart {
  constructor(scope: Construct, id: string, props?: ChartProps) {
    super(scope, id, props);

    const myTask = new PipelineTaskBuilder()
      .withName('fetch-source')
      .withTaskReference('git-clone')
      .withWorkspace('output', 'shared-files', 'The files cloned by the task')
      .withStringParam('url', 'repo-url');

    const pipeline = new Pipeline(this, 'my-pipeline', {
      name: 'my-pipleine',
    });
    pipeline.addTask(myTask);
  }
}

describe('Placeholder', () => {
  test('Empty', () => {
    const app = Testing.app();
    const chart = new MyTestChart(app, 'test-chart');
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
