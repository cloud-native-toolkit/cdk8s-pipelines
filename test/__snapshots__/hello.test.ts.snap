// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder Empty 1`] = `
[
  {
    "apiVersion": "tekton.dev/v1beta1",
    "kind": "Pipeline",
    "metadata": {
      "name": "my-pipleine",
    },
    "spec": {
      "params": [
        {
          "default": "",
          "name": "repo-url",
          "type": "string",
        },
      ],
      "tasks": [
        {
          "name": "fetch-source",
          "params": [
            {
              "name": "url",
              "value": "$(params.repo-url)",
            },
          ],
          "taskRef": TaskRef {
            "name": "git-clone",
          },
          "workspaces": [
            {
              "name": "output",
              "workspace": "shared-files",
            },
          ],
        },
      ],
      "workspaces": [
        {
          "description": "The files that are cloned by the task",
          "name": "shared-files",
        },
      ],
    },
  },
]
`;
